desc/compute/_stability.py

```diff
@@ -231,3 +231,513 @@ def _magnetic_well(params, transforms, profiles, data, **kwargs):
+    provided in the list ``data`` above. Here's a description of
+    these parameters:
+
+    - a: minor radius of the device
+    - g^aa: |grad alpha|^2, field line bending term
+    - g^ra: (grad alpha dot grad rho) integrated local shear
+    - g^rr: |grad rho|^2 flux expansion term
+    - cvdrift: geometric factor of the curvature drift
+    - cvdrift0: geoetric factor of curvature drift 2
+    - |B|: magnitude of the magnetic field
+    - B^zeta: inverse of the jacobian
+    - p_r: dp/drho, pressure gradient
+    - psi_r: radial gradient of the toroidal flux
+    - phi: coordinate describing the position in the toroidal angle
+    along a field line
+
+    Here's how we define the Newcomb metric:
+    If zero crossing is at -inf (root finder failed), use the Y coordinate
+    as a metric of stability else use the zero-crossing point on the X-axis
+    as the metric
+    """
+    rho = data["rho"]
+
+    psi_b = params["Psi"] / (2 * jnp.pi)
+    a_N = data["a"]
+    # a_N is 0.4
+    B_N = 2 * psi_b / a_N**2
+
+    N_zeta0 = int(11)
+    # up-down symmetric equilibria only
+    zeta0 = jnp.linspace(-jnp.pi / 2, jnp.pi / 2, N_zeta0)
+
+    iota = data["iota"]
+    shat = -rho / iota * data["iota_r"]
+    psi = data["psi"]
+    sign_psi = jnp.sign(psi[-1])
+    sign_iota = jnp.sign(iota[-1])
+
+    phi = data["phi"]
+
+    # Count the number of 0s in phi (= number of field lines)
+    N_alpha = int(8)
+    N_zeta = int(len(phi) / N_alpha)
+
+    B = jnp.reshape(data["|B|"], (N_alpha, 1, N_zeta))
+    gradpar = jnp.reshape(data["B^zeta"] / data["|B|"], (N_alpha, 1, N_zeta))
+    dpdpsi = jnp.mean(mu_0 * data["p_r"] / data["psi_r"])
+
+    gds2 = jnp.reshape(
+        rho**2
+        * (
+            data["g^aa"][None, :]
+            - 2 * sign_iota * shat / rho * zeta0[:, None] * data["g^ra"][None, :]
+            + zeta0[:, None] ** 2 * (shat / rho) ** 2 * data["g^rr"][None, :]
+        ),
+        (N_alpha, N_zeta0, N_zeta),
+    )
+
+    g = a_N**3 * B_N * gds2 / B * gradpar
+    g_half = (g[:, :, 1:] + g[:, :, :-1]) / 2
+    c = (
+        1
+        * a_N**3
+        * B_N
+        * jnp.reshape(
+            2
+            / data["B^zeta"][None, :]
+            * sign_psi
+            * rho**2
+            * dpdpsi
+            * (
+                data["cvdrift"][None, :]
+                - shat / rho * zeta0[:, None] * data["cvdrift0"][None, :]
+            ),
+            (N_alpha, N_zeta0, N_zeta),
+        )
+    )
+
+    h = phi[1] - phi[0]
+
+    # g_half on half grid points, c_full on full grid points
+    g_half = (g[:, :, 1:] + g[:, :, :-1]) / 2
+    c_full = c[:, :, :-1]
+
+    i = jnp.arange(N_alpha)[:, None, None]
+    j = jnp.arange(N_zeta0)[None, :, None]
+    k = jnp.arange(N_zeta - 1)[None, None, :]
+
+    X = jnp.zeros((N_alpha, N_zeta0, N_zeta - 1))
+    X = X.at[i, j, k].set(phi[k])
+
+    Y = jnp.zeros((N_alpha, N_zeta0))
+    eps = 5e-3  # slope of the test function
```

Comment:
how was this value decided? can this be changed? does it affect results?

Reply:
It does not, as long as the range of zeta and the resolution in zeta is sufficient. You also probably choose a slope for the test function when you shoot it from the axis in RDCON. Same thing.